src/combinatorics/graph/basic.lean

```diff
@@ -0,0 +1,620 @@
+@[ext] structure weight_simple_graph (α : Type*) (W : Type*) extends weight_simple_digraph α W :=
+(weight_comm (a b : α) : weight a b = weight b a)
+
+/-- Reinterprets a weighted simple graph as a simple graph. -/
+def weight_simple_graph.to_weight_graph (G : weight_simple_graph α W) : weight_graph α W := { .. G }
+
+/-- `weight_simple_graph_class F α β W` states that `F` is a type of `W`-weighted simple graphs on
+`α`.
+
+Extend this class when you extend `weight_simple_graph`. -/
+class weight_simple_graph_class (F : Type*) (α W : out_param $ Type*)
+  extends weight_simple_digraph_class F α W :=
+(weight_comm (f : F) (a b : α) : weight f a b = weight f b a)
+
+@[priority 100] -- see Note [lower instance priority]
+instance weight_simple_graph_class.to_weight_graph_class [weight_simple_graph_class F α W] :
+  weight_graph_class F α W :=
+{ .. ‹weight_simple_graph_class F α W› }
+
+instance weight_simple_graph.weight_simple_graph_class : weight_simple_graph_class (weight_simple_graph α W) α W :=
+{ weight := λ G, G.weight,
+  weight_self := λ G, G.weight_self,
+  weight_comm := weight_simple_graph.weight_comm }
+
+@[priority 100] -- see Note [lower instance priority]
+instance weight_simple_graph_class.to_simple_graph_class (F α W : Type*) [weight_simple_graph_class F α W] :
+  simple_graph_class F α :=
+{ ..weight_simple_digraph_class.to_simple_digraph_class F α W  }
+
+instance : has_bot (weight_simple_graph α W) :=
+⟨{ weight := λ _ _, none,
+  weight_self := λ _, rfl,
+  weight_comm := λ _ _, rfl  }⟩
+
+instance : inhabited (weight_simple_graph α W) := ⟨⊥⟩
+
+
+/-! ### Sort world -/
+
+-- insert `quiver` definition here
+-- Note, `quiver` itself is a class.
+-- Note, we could generalize `quiver α` to `quiver α β`
+
+/-- `quiver_class F α` states that `F` is a type of quivers on `α`.
+
+Extend this class when you extend `quiver`. -/
+class quiver_class (F : Type*) (α : out_param $ Type*) :=
+(hom : F → α → α → Sort*)
+
+instance quiver.quiver_class : quiver_class (quiver α) α :=
+⟨@quiver.hom α⟩
+
+/-- `a ⟶[G] b` is the type of edges from `a` to `b` in `G`. -/
+@[nolint has_inhabited_instance] def graph.hom [quiver_class F α] : F → α → α → Sort* := quiver_class.hom
+
+notation a ` ⟶[`:25 G `] ` b := graph.hom G a b
+
+-- We might need to replace `u₃` by `u₃ + 1` here to avoid providing an adjacency relation for stuff
+-- that already has one, ie when `hom` is already Prop-valued
+@[priority 100] -- see Note [lower instance priority]
+instance quiver_class.to_rel_class (F : Type u₁) (α : Type u₂) [quiver_class.{u₁ u₂ u₃} F α] :
+  rel_class F α α :=
+⟨λ f a b, nonempty (a ⟶[f] b)⟩
+
+instance : has_bot (quiver α) := ⟨⟨λ _ _, pempty⟩⟩
+instance : inhabited (quiver α) := ⟨⊥⟩
+
+
+/-- The type of multigraphs on `α`. -/
+structure multigraph (α : Type u₂) extends quiver.{u₁ u₂} α : Type (max u₁ u₂) :=
+(hom_fun (a b : α) : (a ⟶ b) → (b ⟶ a))
+(hom_fun_comp (a b : α) : hom_fun a b ∘ hom_fun b a = id)
```

Comment:
I think it would probably be more natural to have a single type of edges, i.e. something like that:
```
structure multigraph :=
(V : Type u₁)
(E : Type u₂)
(fst : E → V)
(snd : E → V)
(inv : E → E)
(inv_fst : ∀ (e : E), fst (inv e) = snd e)
(inv_snd : ∀ (e : E), snd (inv e) = fst e)
(inv_inv : ∀ (e : E), inv (inv e) = e)```
That is because we often want to talk about the type of all edges at once, and it allows giving graph morphisms a `has_coe_to_fun` instance making them functions on edges.

Reply:
The difference is that the vertex set is now part of the structure rather than an argument, but I experimented a little bit with it and think it makes sense.